src/Microsoft.TemplateEngine.Edge/Settings/GlobalSettings.cs

```diff
@@ -1,42 +1,230 @@
+            if (!_locked)
+                ReloadSettings(true, null);
+        }
 
-        [JsonProperty]
-        public string DefaultLanguage { get; set; }
+        public event Action? SettingsChanged;
+
+        private List<TemplatesSourceData> _userInstalledTemplatesSources = new List<TemplatesSourceData>();
 
-        [JsonIgnore]
         public IReadOnlyList<TemplatesSourceData> UserInstalledTemplatesSources => _userInstalledTemplatesSources;
 
-        // What this does, if older TemplateEngine loads this file and save it back
-        // it will include new settings that new TemplateEngine depends on
-        // without this field, data would be lost in process of loading and saving
-        //[JsonExtensionData]
-        //private IDictionary<string, JToken> _additionalData;
         public void Add(TemplatesSourceData userInstalledTemplate)
         {
+            if (!_locked)
+                throw new InvalidOperationException($"Call {nameof(LockAsync)} before calling this method");
             _userInstalledTemplatesSources.RemoveAll(data => data.MountPointUri == userInstalledTemplate.MountPointUri);
             _userInstalledTemplatesSources.Add(userInstalledTemplate);
             SettingsChanged?.Invoke();
         }
 
         public void Remove(TemplatesSourceData userInstalledTemplate)
         {
+            if (!_locked)
+                throw new InvalidOperationException($"Call {nameof(LockAsync)} before calling this method");
             _userInstalledTemplatesSources.RemoveAll(data => data.MountPointUri == userInstalledTemplate.MountPointUri);
             SettingsChanged?.Invoke();
         }
+
+        class MyMutex
+        {
+            TaskCompletionSource<bool> _taskCompletionSource;
+            CancellationToken _token;
+            ManualResetEvent _mre = new ManualResetEvent(false);
+
+            public MyMutex(CancellationToken token)
+            {
+                _token = token;
+                _taskCompletionSource = new TaskCompletionSource<bool>();
+                var thread = new Thread(new ThreadStart(WaitLoop));
+                thread.Start();
+            }
+
+            public Task WaitAsync()
+            {
+                return _taskCompletionSource.Task;
+            }
+
+            private void WaitLoop()
+            {
+                var mutex = new Mutex(false, "{01B5E8B0-EF76-48ED-BE95-A0458D7DA2C2}");
+                while (true)
+                {
+                    if (_token.IsCancellationRequested)
+                        return;
+                    if (mutex.WaitOne(20))
+                        break;
+                }
+                _taskCompletionSource.SetResult(true);
+                _mre.WaitOne();
+                mutex.ReleaseMutex();
+            }
+            public void Release()
+            {
+                _mre.Set();
+            }
+        }
+
+        class DisposableCallback : IDisposable
```

Comment:
This is a non-standard pattern. Why not do both resource disposal and resource acquisition in this IDisposable class? In other words, please consider moving the logic from `LockAsync` to a constructor or factory method on this class and eliminate the delegate. And then maybe call it `LockHolder` or something similar.

`LockAsync` then simply becomes something like:

```C#
  return new LockHolder(this, token);
```

Reply:
I believe I followed this suggestion in updated PR.